### PR TITLE
[mesheryctl] Added `pattern view`

### LINKED ISSUE #2624: [mesheryctl] `pattern view`
Refer the [mesheryctl Command Tracker](https://docs.google.com/spreadsheets/d/1q63sIGAuCnIeDs8PeM-0BAkNj8BBgPUXhLbe1Y-318o/edit#gid=0) for current status of `mesheryctl`.

#### Current Behavior
Related to #2266 and #2621, `mesheryctl` v0.5.0 supports the following `exp pattern` command operations:

```
Available Commands:
  apply       Apply pattern file
  delete      Delete pattern file
```

Meshery release v0.6.0 will support `pattern` as a stable command with `view` being one of the subcommands.

#### Desired Behavior
The following subcommands and flags should be supported as well:

```
  view       Displays the contents of a specific pattern file

Flags:
   -a, --all    Display the contents of all available pattern files
   -o json     Displays the contents of the pattern(s) in json format
```

### Implementation
1. Use `http://<current-context-endpoint>/api/experimental/patternfile` to retrieve a list of available patterns.
1. Require use of token to auth to endpoint.
1. Display the contents of the pattern(s) in yaml format with optional `-a`, `--all` flag to show all pattern file content.
1. Display the contents of the pattern(s) in json format with optional `-o json` flag.
1. Do not show total count of available patterns.

#### Acceptance Tests
1. `pattern view [pattern-name|pattern-id]` command should display the contents of the requested pattern. Patterns may be identified by name or by id.
2. `pattern view [pattern-name|pattern-id]` will require `-t` or `--token` has a valid token file as a flag.
3. `pattern view [pattern-name|pattern-id]` will display the contents of the pattern in yaml format by default. See mockup.
4. `pattern view [pattern-name|pattern-id]-o json` will display the contents of the pattern in yaml format by default. See mockup.
5. `pattern view -a` will display the contents of all available patterns in yaml format by default. See mockup.
6. `pattern view -a -o json` will display the contents of all available patterns in json format. See mockup.

### Mockups
`view -a` output:

```
patterns:
- id: 2a634cfe-87d5-478e-9fda-168580ff83ba
  name: canary-rollout.yaml
  user_id: '090e7114-509a-4046-81f1-9c5fb8daf724'
  pattern_file: "name: IstioSM\nservices:\n  istio:\n    type: Pod\n    namespace:
    istio-system\n    settings:\n      version: 1.8.2\n    traits:\n      mTLS:\n
    \       policy: mutual\n        namespaces:\n          - istio-test\n      automaticSidecarInjection:\n
    \       namespaces:\n          - default\n          - istio-test\n      meshmap:\n
    \       position:\n          posX: 98\n          posY: 54.8   \n          \n  istio2:\n
    \  type: Service\n   namespace: LinkerD-system\n   settings:\n     version: 1.8.2\n
    \  traits:\n     mTLS:\n       policy: mutual\n       namespaces:\n         -
    istio-test\n     automaticSidecarInjection:\n       namespaces:\n         - default\n
    \        - istio-test\n     meshmap:\n       position:\n         posX: 298\n         posY:
    54.8\n"
  created_at: '2021-03-23T20:28:50.229379Z'
  updated_at: '2021-03-23T20:28:50.229389Z'
- id: 3850572d-c7b5-4260-8c06-a45932a7b658
  name: pattern.yaml
  user_id: 6e955e6d-7a8f-4687-b0bb-17504aa8d520
  pattern_file: "name: IstioSM\nservices:\n  istio:\n    type: Pod\n    namespace:
    istio-system\n    settings:\n      version: 1.8.2\n    traits:\n      mTLS:\n
    \       policy: mutual\n        namespaces:\n          - istio-test\n      automaticSidecarInjection:\n
    \       namespaces:\n          - default\n          - istio-test\n      meshmap:\n
    \       position:\n          posX: 98\n          posY: 54.8   \n          \n  istio2:\n
    \  type: Service\n   namespace: LinkerD-system\n   settings:\n     version: 1.8.2\n
    \  traits:\n     mTLS:\n       policy: mutual\n       namespaces:\n         -
    istio-test\n     automaticSidecarInjection:\n       namespaces:\n         - default\n
    \        - istio-test\n     meshmap:\n       position:\n         posX: 298\n         posY:
    54.8"
  created_at: '2021-03-08T18:11:24.874038Z'
  updated_at: '2021-03-23T05:33:34.357108Z'
```

`view -a -o json` output:
```
  "patterns": [
    {
      "id": "2a634cfe-87d5-478e-9fda-168580ff83ba",
      "name": "canary-rollout.yaml",
      "user_id": "090e7114-509a-4046-81f1-9c5fb8daf724",
      "pattern_file": "name: IstioSM\nservices:\n  istio:\n    type: Pod\n    namespace: istio-system\n    settings:\n      version: 1.8.2\n    traits:\n      mTLS:\n        policy: mutual\n        namespaces:\n          - istio-test\n      automaticSidecarInjection:\n        namespaces:\n          - default\n          - istio-test\n      meshmap:\n        position:\n          posX: 98\n          posY: 54.8   \n          \n  istio2:\n   type: Service\n   namespace: LinkerD-system\n   settings:\n     version: 1.8.2\n   traits:\n     mTLS:\n       policy: mutual\n       namespaces:\n         - istio-test\n     automaticSidecarInjection:\n       namespaces:\n         - default\n         - istio-test\n     meshmap:\n       position:\n         posX: 298\n         posY: 54.8\n",
      "created_at": "2021-03-23T20:28:50.229379Z",
      "updated_at": "2021-03-23T20:28:50.229389Z"
    },
    {
      "id": "3850572d-c7b5-4260-8c06-a45932a7b658",
      "name": "pattern.yaml",
      "user_id": "6e955e6d-7a8f-4687-b0bb-17504aa8d520",
      "pattern_file": "name: IstioSM\nservices:\n  istio:\n    type: Pod\n    namespace: istio-system\n    settings:\n      version: 1.8.2\n    traits:\n      mTLS:\n        policy: mutual\n        namespaces:\n          - istio-test\n      automaticSidecarInjection:\n        namespaces:\n          - default\n          - istio-test\n      meshmap:\n        position:\n          posX: 98\n          posY: 54.8   \n          \n  istio2:\n   type: Service\n   namespace: LinkerD-system\n   settings:\n     version: 1.8.2\n   traits:\n     mTLS:\n       policy: mutual\n       namespaces:\n         - istio-test\n     automaticSidecarInjection:\n       namespaces:\n         - default\n         - istio-test\n     meshmap:\n       position:\n         posX: 298\n         posY: 54.8",
      "created_at": "2021-03-08T18:11:24.874038Z",
      "updated_at": "2021-03-23T05:33:34.357108Z"
    }
  ]
}
```

---

#### Contributor Resources

- [mesheryctl Contributing Guide](https://github.com/layer5io/meshery/blob/master/mesheryctl/README.md)
- [Beginner's guide to contributing to Meshery and mesheryctl](https://youtu.be/hh_kFLZx3G4)
- [mesheryctl Command Tracker](https://docs.google.com/spreadsheets/d/1q63sIGAuCnIeDs8PeM-0BAkNj8BBgPUXhLbe1Y-318o/edit#gid=0)
- [Meshery CLI Commands and Documentation](https://docs.google.com/document/d/1xRlFpElRmybJ3WacgPKXgCSiQ2poJl3iCCV1dAalf0k/edit#heading=h.5fucij4hc5wt)

### PR DESCRIPTION
**Pull request by aneeshnema**
**Description**

This PR fixes #2624 

**Notes for Reviewers**

- [x] view pattern by name or id
- [x] json, yaml(default) available as output formats
- [x] view all patterns
- [x] pretty print the patterns (indentation)
- [x] token flag

**[Signed commits](https://github.com/layer5io/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by aneeshnema** [comment]
Regarding the token flag, should this be added only for the view command or as a persistent to the pattern (Since the other sub-commands also need token)

**Comment by leecalcote** [comment]
@aneeshnema as a requirement for all pattern commands, yes. In the future it may be that some patterns are generally available to all users of a Meshery system, but even then, the user would need to auth to Meshery Server in order to list, view, apply, delete them.

**Comment by leecalcote** [comment]
Sheesh. @aneeshnema, you tossing this up faster than we can review them. 💪

**Comment by navendu-pottekkat** [comment]
@leecalcote @aneeshnema should this PR be modified from our latest discussion on the development call? I mean the ID vs name we talked about.

File: mesheryctl/internal/cli/root/pattern/view.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+		mctlCfg, err := config.GetMesheryCtl(viper.GetViper())
+		if err != nil {
+			return errors.Wrap(err, "error processing config")
+		}
+		pattern := ""
+		isID := false
+		// if pattern name/id available
+		if len(args) > 0 {
+			if viewAllFlag {
+				return errors.New("-a cannot be used when pattern-name(pattern-id) is specified")
```
Comment (navendu-pottekkat) [inline mesheryctl/internal/cli/root/pattern/view.go:38]: `[pattern-name | pattern-id]` would be the correct usage.

File: mesheryctl/internal/cli/root/pattern/view.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+			return errors.New("output-format choice invalid, use json or yaml")
+		}
+		log.Info(string(body))
+		return nil
+	},
+}
+
+func init() {
+	viewCmd.Flags().BoolVarP(&viewAllFlag, "all", "a", false, "(optional) view all patterns available")
+	viewCmd.Flags().StringVarP(&outFormatFlag, "output-format", "o", "yaml", "(optional) format to display in (json|yaml)")
```
Comment (navendu-pottekkat) [inline mesheryctl/internal/cli/root/pattern/view.go:128]: `[json|yaml]`

**Review by navendu-pottekkat — COMMENTED** [review]
@aneeshnema left some changes needed for consistency. We also need to make the `--token` flag required and mention that on the usage as the command would not work otherwise.

File: mesheryctl/internal/cli/root/pattern/view.go (hunk 1 above)
Comment (navendu-pottekkat) [reply mesheryctl/internal/cli/root/pattern/view.go:38]: Same change for the other occurrences.

**Comment by aneeshnema** [comment]
@navendu-pottekkat, currently, the pattern command also tries to get the token path from the environment variable `MESHERY_AUTH_TOKEN`, making the token flag compulsory will stop users from using the environment variable way to set token path. (decreasing the ease of use)

// @leecalcote

File: mesheryctl/internal/cli/root/pattern/view.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+	"github.com/spf13/viper"
+)
+
+var (
+	viewAllFlag   bool
+	outFormatFlag string
+)
+
+var viewCmd = &cobra.Command{
+	Use:   "view",
```
Comment (navendu-pottekkat) [inline mesheryctl/internal/cli/root/pattern/view.go:24]:
```suggestion
	Use:   "view <pattern name>",
```

**Review by navendu-pottekkat — COMMENTED** [review]
Looks good @aneeshnema .

One small concern is that the usage shown is `mesheryctl pattern view [flags]`.

Shouldn't it be `mesheryctl pattern view <pattern name> [flags]`?

I have marked the change below- please check it out.

**Review by navendu-pottekkat — APPROVED** [review]
LGTM 💯

**Pull request merged**